### DISCUSSION TITLE
Move Thrift dependency to FBThrift

### DISCUSSION THREAD

**Discussion by pedroerp** [facebookincubator/velox#6163] [General]
We currently have a dependency on thrift from our native parquet reader. As we add support for remote function execution, the client <-> server communication will be done using thrift RPC as well. 

FBThrift is a newer and supposedly more efficient version than thrift, and among other things provides an async transport called "cpp2" (not available in thrift), which will be used for remote function execution. 

Linking thrift and fbthrift to the same binary causes symbols to conflict and end up causing crashes as the wrong version of symbols eventually get picked. There was a long discussion on Slack (check #builds-and-ci channel) on whether this could be solved using dynamic linking, but so solution was found. 

With this context, the proposal is to remove the thrift dependency from Velox (created due to parquet), and consolidate all of that on FBThrift. Despite their divergence, both projects seem to support TCompactProtocol (the ones used by Parquet), so, in theory (need some testing to validate), they should be compatible. FBThrift is also claimed to be more efficient, so there might be small performance gains in doing so.

The parquet generated cpp files are also today pre-compiled and checked in as part of the codebase; as part of this change we would also move them to be compiled at build time, like how it is done for remote functions. I've tested and all of them seems to be doable. 

There is also a thrift dependency that comes with Arrow (we use arrow's parquet writer), but this is nicely wrapped within arrow and doesn't leak its internal symbols. 

What do you all think?

UPDATE: just doubled checked with someone from the fbthrift team and the TCompactProtocol serialization needed for Parquet should be compatible. 

Cc: @assignUser @majetideepak @aditi-pandit @yingsu00 @kgpai @mbasmanova @xiaoxmeng @oerling

**Comment by assignUser** [comment]
Makes sense to not support/use both versions of thrift and consolidate.

**Comment by majetideepak** [comment]
+1.  We have faced thrift/fbthrift conflict in the past as well with Prestissimo. See https://github.com/prestodb/presto/issues/20340 and https://github.com/prestodb/presto/issues/18261. 
We should have only one of them. Choosing FBThrift makes sense given it has better capabilities.

**Reply by assignUser** [reply]
I just had a look at fbthrift and they use the `apache::thrift` namespace... https://github.com/facebook/fbthrift/blob/5c8e8b8d449061e446f28a97370995113ddf48c4/thrift/compiler/generate/common.cc#L29

No wonder symbols are clashing...

**Reply by majetideepak** [reply]
Likely the segfaults and runtime errors are ODR violations.

**Reply by majetideepak** [reply]
We are seeing runtime errors with the Velox static mono lib as well.
https://github.com/prestodb/presto/actions/runs/12696402195/job/35393520568?pr=24343

**Comment by majetideepak** [comment]
@pedroerp can we please get some performance validation for Parquet Reader with this switch and ensure there are no surprises? Thanks!

**Reply by pedroerp** [reply]
Good point. Other than potentially micro-benchmarks, do you guys have other ways to test parquet performance? Internally we only use the DWRF reader.

**Reply by majetideepak** [reply]
There is the ParquetReaderBenchmark in Velox that can be used I believe. @yingsu00 any thoughts on evaluating fbthrift?

**Comment by kevinwilfong** [comment]
Just checking, was there a conclusion here?  It seems like the discussion fizzled out and the PR to make the change went stale https://github.com/facebookincubator/velox/pull/6523

**Reply by kevinwilfong** [reply]
Asking because it looks we ran into issues again related to using thrift instead of fbthrift https://github.com/facebookincubator/velox/pull/11728

**Reply by majetideepak** [reply]
@kevinwilfong We all agree that we should remove the thrift dependency in favor of fbthrift.

**Reply by majetideepak** [reply]
The challenge is that Arrow still requires thrift. We don't have a clean way to fully remove thrift.

**Reply by majetideepak** [reply]
Since we disabled Parquet inside Arrow, I realized we won't need thrift.
